### PR TITLE
feat: support --with-access-token for auth

### LINKED ISSUE #406: Implement Token Exchange auth with CLI to enable using encrypt/decrypt as an subprocess operation in unsupported languages
OpenTDF doesn't support some major languages, and it isn't on the roadmap (i.e. Ruby, Python, Rust, etc). Although performance would be much better if we added this support, we don't have capacity to maintain all these SDKs.

By adding token exchange we will enable developers to use the CLI to interact with the CLI as a subprocess and perform encrypt and decrypt ops.

## Acceptance Criteria

- Add support for token exchange
- Add e2e tests
- Add example in Python, Node, and Ruby

### PR DESCRIPTION
**Pull request by jrschumacher**
Closes #406

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
+// if onlyNew is set to true, a new profile will be created and returned
+// returns the profile and the current profile store
+func InitProfile(c *cli.Cli, onlyNew bool) (*profiles.Profile, *profiles.ProfileStore) {
 	var err error
 	profileName := c.FlagHelper.GetOptionalString("profile")
 
 	profile, err = profiles.New()
 	if err != nil || profile == nil {
-		c.ExitWithError("Failed to initialize profile store", err)
+		c.ExitWithError(fmt.Sprintf("Failed to initialize profile store: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:44]:
```suggestion
		c.ExitWithError("Failed to initialize profile store", err)
```

File: cmd/root.go [hunk 2]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
-		c.ExitWithWarning("No default profile set. Use `" + config.AppName + " profile create <profile> <endpoint>` to create a default profile.")
+		c.ExitWithWarning(fmt.Sprintf("No default profile found. Please create one using %s", config.AppName))
 	}
-	c.Printf("Using profile [%s]\n", profile.GetGlobalConfig().GetDefaultProfile())
 
 	if profileName == "" {
 		profileName = profile.GetGlobalConfig().GetDefaultProfile()
 	}
 
+	c.Printf(fmt.Sprintf("Using profile: %s", profileName))
```
Comment (jrschumacher) [inline cmd/root.go:61]:
```suggestion
	c.Printf(fmt.Sprintf("Using profile [%s]\n", profileName))
```

File: cmd/root.go [hunk 3]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
+			cli.ExitWithError(fmt.Sprintf("One of %s must be set", cli.PrettyList(authFlags)), err)
+		} else if authFlagsCounter > 1 {
+			cli.ExitWithError(fmt.Sprintf("Only one of %s must be set", cli.PrettyList(authFlags)), err)
 		}
 
 		inMemoryProfile = true
 		profile, err = profiles.New(profiles.WithInMemoryStore())
 		if err != nil || profile == nil {
-			cli.ExitWithError("Failed to initialize a temporary profile", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to initialize in-memory profile: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:120]:
```suggestion
			cli.ExitWithError("Failed to initialize in-memory profile", err)
```

File: cmd/root.go [hunk 4]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 		}
 
 		if err := profile.AddProfile("temp", host, tlsNoVerify, true); err != nil {
-			cli.ExitWithError("Failed to create temporary profile", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to create in-memory profile: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:124]:
```suggestion
			cli.ExitWithError("Failed to create in-memory profile", err)
```

File: cmd/root.go [hunk 5]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 		}
 
 		// add credentials to the temporary profile
 		cp, err = profile.UseProfile("temp")
 		if err != nil {
-			cli.ExitWithError("Failed to load temporary profile", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to load in-memory profile: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:130]:
```suggestion
			cli.ExitWithError("Failed to load in-memory profile", err)
```

File: cmd/root.go [hunk 6]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
-			AuthType:     profiles.PROFILE_AUTH_TYPE_CLIENT_CREDENTIALS,
-			ClientId:     cc.ClientId,
-			ClientSecret: cc.ClientSecret,
-		}); err != nil {
-			cli.ExitWithError("Failed to set client credentials", err)
+		// get credentials from flags
+		if withAccessToken != "" {
+			claims, err := auth.ParseClaimsJWT(withAccessToken)
+			if err != nil {
+				cli.ExitWithError(fmt.Sprintf("Failed to get access token: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:137]:
```suggestion
				cli.ExitWithError("Failed to get access token", err)
```

File: cmd/root.go [hunk 7]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
+			}
+
+			if err := cp.SetAuthCredentials(profiles.AuthCredentials{
+				AuthType: profiles.PROFILE_AUTH_TYPE_ACCESS_TOKEN,
+				AccessToken: profiles.AuthCredentialsAccessToken{
+					AccessToken: withAccessToken,
+					Expiration:  claims.Expiration,
+				},
+			}); err != nil {
+				cli.ExitWithError(fmt.Sprintf("Failed to set access token: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:147]:
```suggestion
				cli.ExitWithError("Failed to set access token", err)
```

File: cmd/root.go [hunk 8]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
+				cli.ExitWithError(fmt.Sprintf("Failed to get client credentials: %v", err), err)
+			}
+
+			// add credentials to the temporary profile
+			if err := cp.SetAuthCredentials(profiles.AuthCredentials{
+				AuthType:     profiles.PROFILE_AUTH_TYPE_CLIENT_CREDENTIALS,
+				ClientId:     cc.ClientId,
+				ClientSecret: cc.ClientSecret,
+			}); err != nil {
+				cli.ExitWithError(fmt.Sprintf("Failed to set client credentials: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:166]:
```suggestion
				cli.ExitWithError("Failed to set client credentials", err)
```

File: cmd/root.go [hunk 9]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
+			}
 		}
 		if err := cp.Save(); err != nil {
-			cli.ExitWithError("Failed to save profile", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to save profile: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:170]:
```suggestion
			cli.ExitWithError("Failed to save profile", err)
```

File: cmd/root.go [hunk 10]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 	}
 
 	if err := auth.ValidateProfileAuthCredentials(c.Context(), cp); err != nil {
 		if errors.Is(err, auth.ErrPlatformConfigNotFound) {
-			cli.ExitWithError(fmt.Sprintf("Failed to get platform configuration. Is the platform accepting connections at '%s'?", cp.GetEndpoint()), nil)
+			cli.ExitWithError(fmt.Sprintf("Platform configuration not found for endpoint: %s", cp.GetEndpoint()), nil)
 		}
 		if inMemoryProfile {
-			cli.ExitWithError("Failed to authenticate with flag-provided client credentials", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to authenticate: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:181]:
```suggestion
			cli.ExitWithError("Failed to authenticate with flag-provided client credentials.", err)
```

File: cmd/root.go [hunk 11]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 		}
 	} else {
-		cp = InitProfile(c, false)
+		profile, cp = InitProfile(c, false)
 	}
 
 	if err := auth.ValidateProfileAuthCredentials(c.Context(), cp); err != nil {
 		if errors.Is(err, auth.ErrPlatformConfigNotFound) {
-			cli.ExitWithError(fmt.Sprintf("Failed to get platform configuration. Is the platform accepting connections at '%s'?", cp.GetEndpoint()), nil)
+			cli.ExitWithError(fmt.Sprintf("Platform configuration not found for endpoint: %s", cp.GetEndpoint()), nil)
```
Comment (jrschumacher) [inline cmd/root.go:178]:
```suggestion
			cli.ExitWithError(fmt.Sprintf("Failed to get platform configuration. Is the platform accepting connections at '%s'?", cp.GetEndpoint()), nil)
```

File: cmd/root.go [hunk 12]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 		}
 		if inMemoryProfile {
-			cli.ExitWithError("Failed to authenticate with flag-provided client credentials", err)
+			cli.ExitWithError(fmt.Sprintf("Failed to authenticate: %v", err), err)
 		}
 		if errors.Is(err, auth.ErrProfileCredentialsNotFound) {
-			cli.ExitWithWarning("Profile missing credentials. Please login or add client credentials.")
+			cli.ExitWithWarning("Profile missing credentials")
```
Comment (jrschumacher) [inline cmd/root.go:184]:
```suggestion
			cli.ExitWithWarning("Profile missing credentials. Please login or add client credentials.)
```

File: cmd/root.go [hunk 13]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 		if errors.Is(err, auth.ErrAccessTokenExpired) {
-			cli.ExitWithWarning("Access token expired. Please login again.")
+			cli.ExitWithWarning("Access token expired. Please login or add flag-provided credentials.")
 		}
 		if errors.Is(err, auth.ErrAccessTokenNotFound) {
-			cli.ExitWithWarning("No access token found. Please login or add client credentials.")
+			cli.ExitWithWarning("No access token found. Please login or add flag-provided credentials.")
 		}
-		cli.ExitWithError("Failed to get access token", err)
+		cli.ExitWithError(fmt.Sprintf("Failed to get access token: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:193]:
```suggestion
		cli.ExitWithError("Failed to get access token.", err)
```

File: cmd/root.go [hunk 14]
```diff
@@ -35,136 +32,170 @@ type version struct {
...
 	}
 
 	h, err := handlers.New(handlers.WithProfile(cp))
 	if err != nil {
-		cli.ExitWithError("Failed to create handler", err)
+		cli.ExitWithError(fmt.Sprintf("Unexpected error: %v", err), err)
```
Comment (jrschumacher) [inline cmd/root.go:198]:
```suggestion
		cli.ExitWithError("Unexpected error", err)
```

File: docs/man/_index.md [hunk 15]
```diff
@@ -26,6 +26,8 @@ command:
         - fatal
         - panic
       default: info
+    - name: with-access-token
+      description: access token for authentication via bearer token
```
Comment (jakedoublev) [inline docs/man/_index.md:30]: Is it a requirement that this access-token has been granted from the same client as the running platform's `public_client_id` client, or is it any AT from the idP running alongside the platform?

Comment (jakedoublev) [reply docs/man/_index.md:30]: I'm guessing the latter, but we might need to test and document in this manual either way.

File: pkg/auth/auth.go [hunk 16]
```diff
@@ -141,6 +136,20 @@ func buildToken(c *profiles.AuthCredentials) *oauth2.Token {
 	}
 }
 
+func ParseClaimsJWT(accessToken string) (JWTClaims, error) {
+	c := struct {
+		Expiration int64 `json:"exp"`
+	}{}
```
Comment (jakedoublev) [inline pkg/auth/auth.go:142]:
```suggestion
	c := JWTClaims{}
```

File: docs/man/_index.md (hunk 15 above)
Comment (jakedoublev) [reply docs/man/_index.md:30]: Validated locally with an access token granted from the client credentials flow (a client creds profile, `auth print-access-token`, and pasting the token into a `--with-access-token` request to list attributes).

**Review by jakedoublev — DISMISSED** [review]

File: cmd/root.go [hunk 17]
```diff
@@ -45,126 +45,157 @@ func InitProfile(c *cli.Cli, onlyNew bool) *profiles.ProfileStore {
...
-		c.ExitWithWarning("No default profile set. Use `" + config.AppName + " profile create <profile> <endpoint>` to create a default profile.")
+		c.ExitWithWarning(fmt.Sprintf("No default profile set. Use `%s profile create <profile> <endpoint>` to create a default profile.", config.AppName))
 	}
-	c.Printf("Using profile [%s]\n", profile.GetGlobalConfig().GetDefaultProfile())
 
 	if profileName == "" {
 		profileName = profile.GetGlobalConfig().GetDefaultProfile()
 	}
 
+	c.Printf(fmt.Sprintf("Using profile [%s]\n", profileName))
```
Comment (jrschumacher) [inline cmd/root.go:61]:
```suggestion
	c.Printf("Using profile [%s]\n", profileName)
```

**Review by jakedoublev — APPROVED** [review]

**Pull request merged**